src/editor/codemirror/edits.ts

```diff
@@ -153,6 +156,61 @@ export const calculateChanges = (
   });
 };
 
+const findIndentLevel = (
+  state: EditorState,
+  mainFrom: number,
+  levelHint: number | undefined
+): number => {
+  // This affects whether we can lower the indent level based on the hint.
+  let nextNonBlankLineIndentLevel: number = 0;
+  for (const line of succeedingLinesInclusive(state, mainFrom)) {
+    const text = line.text;
+    if (text.trim()) {
+      nextNonBlankLineIndentLevel = indentLevel(text);
+      break;
+    }
+  }
+  // Now base our indent level on the preceeding non-blank line, using the hint
+  // if it will reduce indentation and we're not mid block.
+  for (const line of preceedingLinesExclusive(state, mainFrom)) {
+    const text = line.text;
+    if (text.trim()) {
+      let indent = indentLevel(text);
+      // Beginning of block:
+      if (text.trim().endsWith(":")) {
+        indent += 1;
+      } else if (
+        levelHint !== undefined &&
+        levelHint < indent &&
+        nextNonBlankLineIndentLevel < indent &&
```

Comment:
We also shouldn't be able to go back further than nextNonBlankLineIndentLevel. So we need the max of that and the hint.

Reply:
I didn't express is quite right above. Clearer in code: e865038d866e5eb386117e01fe03422e889ac033